dataactvalidator/scripts/load_program_activity.py

```diff
@@ -20,54 +29,179 @@ def get_program_activity_file(base_path):
+    logger.info('Loading program activity: ' + PA_FILE_NAME)
 
     with create_app().app_context():
         sess = GlobalDB.db().session
+        try:
+            data = pd.read_csv(program_activity_file, dtype=str)
+        except pd.io.common.EmptyDataError as e:
+            log_blank_file()
+            sys.exit(4)  # exit code chosen arbitrarily, to indicate distinct failure states. Approved by Ops.
+        try:
+            dropped_count, data = clean_data(
+                data,
+                ProgramActivity,
+                {"fyq": "fiscal_year_quarter", "agency_id": "agency_id", "allocation_id": "allocation_transfer_id",
+                 "account": "account_number", "pa_code": "program_activity_code", "pa_name": "program_activity_name"},
+                {"program_activity_code": {"pad_to_length": 4}, "agency_id": {"pad_to_length": 3},
+                 "allocation_transfer_id": {"pad_to_length": 3, "keep_null": True},
+                 "account_number": {"pad_to_length": 4}},
+                ["agency_id", "program_activity_code", "account_number", "program_activity_name"],
+                True
+            )
+        except FailureThresholdExceededException as e:
+            if e.count == 0:
+                log_blank_file()
+                sys.exit(4)
+            else:
+                count_str = "Application tried to drop {} rows ".format(e.count)
+                logger.error("Loading of program activity file failed due to exceeded failure threshold. " + count_str)
+                sys.exit(5)
 
         sess.query(ProgramActivity).delete()
 
-        data = pd.read_csv(program_activity_file, dtype=str)
-        data = clean_data(
-            data,
-            ProgramActivity,
-            {"fyq": "fiscal_year_quarter", "agency_id": "agency_id", "alloc_id": "allocation_transfer_id",
-             "account": "account_number", "pa_code": "program_activity_code", "pa_name": "program_activity_name"},
-            {"program_activity_code": {"pad_to_length": 4}, "agency_id": {"pad_to_length": 3},
-             "allocation_transfer_id": {"pad_to_length": 3, "keep_null": True}, "account_number": {"pad_to_length": 4}}
-        )
         # Lowercase Program Activity Name
-        data['program_activity_name'] = data['program_activity_name'].apply(lambda x: x.lower())
+        data['program_activity_name'] = data['program_activity_name'].apply(lambda x: lowercase_or_notify(x))
+
         # because we're only loading a subset of program activity info,
         # there will be duplicate records in the dataframe. this is ok,
-        # but need to de-duped before the db load.
-        data.drop_duplicates(inplace=True)
+        # but need to de-duped before the db load. We also need to log them.
+
+        deduped = data.drop_duplicates()
+        dropped_dupes = data[np.invert(data.index.isin(deduped.index))]
+        logger.info("Dropped {} duplicate rows.".format(len(dropped_dupes.index)))
+        data = deduped
```

Comment:
For this process, if the only reason we're not dropping inplace is because we want a count, couldn't we just get the original length, then drop dupes and do the original length minus this one? Then we wouldn't have to create two new objects, just a single count. Probably also faster just counting than doing inversions, etc.